libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,236 @@
+      if(in_flight) {
+         auto until = std::chrono::steady_clock::now() + std::chrono::seconds(5);
+         ctx.restart();
+         while(in_flight && ctx.run_one_until(until));
+      }
+
+      if(message_deque.size()) {
+         try {
+            boost::filesystem::create_directories(data_file_path.parent_path());
+
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("wb");
+            fc::raw::pack(file, message_deque);
+         } FC_LOG_AND_DROP();
+      }
+   }
+
+   void retry_connection() {
+      connected = false;
+      channel.reset();
+      connection.reset();
+      handler.reset();
+      next_retry_time = std::min(next_retry_time, 30u);
+      retry_timer.expires_from_now(boost::posix_time::seconds(next_retry_time));
+      retry_timer.async_wait([&](auto ec) {
+         if(ec)
+            return;
+         handler = std::make_unique<reliable_amqp_publisher_handler>(this, ctx);
+         try {
+            connection = std::make_unique<AMQP::TcpConnection>(handler.get(), *amqp_address);
+         } catch(...) { //this should never happen, but it could technically have thrown
+            retry_connection();
+         }
+      });
+      next_retry_time *= 2;
+   }
+
+   void amqp_ready() override {
+      ilog("AMQP connection ${s} publishing to \"${e}\" established", ("s", (std::string)*amqp_address)("e", exchange));
+      next_retry_time = 1;
+      connected = true;
+      channel = std::make_unique<AMQP::TcpChannel>(connection.get());
+      channel->onError([this](const char* s) {
+         elog("Channel error for AMQP connection ${s} publishing to \"${e}\": ${m}; restarting connection", ("s", (std::string)*amqp_address)("e", exchange)("m",s));
+         retry_connection();
+      });
+   }
+
+   void amqp_error(const char* message) override {
+      wlog("AMQP connection ${a} publishing to \"${e}\" failed: ${m}; retrying", ("a", (std::string)*amqp_address)("m", message)("e", exchange));
+      retry_connection();
+   }
+
+   void pump_queue() {
+      if(stopping || in_flight || !connected || message_deque.empty())
+         return;
+
+      constexpr size_t max_msg_single_transaction = 16;
+      const unsigned msgs_this_transaction = std::min(message_deque.size(), max_msg_single_transaction);
+
+      channel->startTransaction();
+      std::for_each_n(message_deque.begin(), msgs_this_transaction, [this](const std::vector<char>& msg) {
+         channel->publish(exchange, "", msg.data(), msg.size());
+      });
+
+      in_flight = msgs_this_transaction;
+
+      channel->commitTransaction().onSuccess([this](){
+         message_deque.erase(message_deque.begin(), message_deque.begin()+in_flight);
+         logged_exceeded_max_depth = false;
+      })
+      .onFinalize([this]() {
+         in_flight = 0;
+         //unfortuately we don't know if an error is due to something recoverable or if an error is due
```

Comment:
#1. If commitTransaction() fails, we may need to log it.  We can get useful information such as how often it happens, the message_id, the number of messages in the envelop, the size of envelop, etc.

#2. Can the size of AMQP::Envelop be too big for rabbitmq if the the number or message with the same message_id is too many and too big?

Reply:
2) The default message size in rabbitmq is 128MB; rabbitmq's max configurable is 512MB; amqp's max is 2GB. These limits were no where close to what I was expecting this to be used for originally. @heifner is rodeos usage a lot of data?

Reply:
3) I think any error is going to bubble up as a channel error or connection error, both of which have a 1s retry timer on them. You can see this by doing actions like publishing to a non-exsisting exchange, or publishing to an exchange that is bound to a queue that is full (and rejecting publishing due to that). Both of those result in a channel error and a 1s retry before attempting again.

Reply:
2. No way they will be anywhere near that size.